wp-includes/rest-api/endpoints/class-wp-rest-customize-partials-controller.php

```diff
@@ -0,0 +1,327 @@
+	 * @access public
+	 *
+	 * @param WP_REST_Request $request Full details about the request.
+	 * @return WP_REST_Response|WP_Error Response object on success, or WP_Error object on failure.
+	 */
+	public function get_item( $request ) {
+		$wp_customize = $this->ensure_customize_manager();
+
+		$wp_customize->selective_refresh->add_dynamic_partials( array( $request['partial'] ) );
+		$partial = $wp_customize->selective_refresh->get_partial( $request['partial'] );
+
+		return rest_ensure_response( $this->prepare_item_for_response( $partial, $request ) );
+	}
+
+	/**
+	 * Checks if a given request has access to read partials.
+	 *
+	 * @since 4.?.?
+	 * @access public
+	 *
+	 * @param  WP_REST_Request $request Full details about the request.
+	 * @return true|WP_Error True if the request has read access, WP_Error object otherwise.
+	 */
+	public function get_items_permissions_check( $request ) {
+		return current_user_can( 'edit_theme_options' );
+	}
+
+	/**
+	 * Retrieves list of partials.
+	 *
+	 * @since 4.?.?
+	 * @access public
+	 *
+	 * @param WP_REST_Request $request Full details about the request.
+	 * @return WP_REST_Response|WP_Error Response object on success, or WP_Error object on failure.
+	 */
+	public function get_items( $request ) {
+		$wp_customize = $this->ensure_customize_manager();
+
+		$all_partials = $wp_customize->selective_refresh->partials();
+		$partials = array();
+
+		foreach ( $all_partials as $partial_id => $partial ) {
+			if ( ! $partial->check_capabilities() ) {
+				continue;
+			}
+			$data = $this->prepare_item_for_response( $partial, $request );
+			$partials[] = $this->prepare_response_for_collection( $data );
+		}
+
+		return rest_ensure_response( $partials );
+	}
+
+	/**
+	 * Prepares a single partial's output for response.
+	 *
+	 * @since 4.?.?
+	 * @access public
+	 *
+	 * @param WP_Customize_Partial $partial WP_Customize_Partial object.
+	 * @param WP_REST_Request      $request Request object.
+	 * @return WP_REST_Response Response object.
+	 */
+	public function prepare_item_for_response( $partial, $request ) {
+
+		$data = array();
+
+		$links = array(
+			'self' => array(
+				'href' => rest_url( trailingslashit( $this->namespace . '/' . $this->rest_base ) . $partial->id ),
+			),
+		);
+
+		$hide_from_response = array(
+			'capability',
+			'component',
+			'manager',
+			'id_data',
+			'primary_setting', // Hide this from response since that's returned within 'settings'.
+			'render_callback',
+		);
+
+		// Get primary setting ID.
+		if ( ! empty( $partial->primary_setting ) ) {
+			$data['settings'] = array( $partial->primary_setting );
+			$links['related'] = array( array(
+				'href' => rest_url( trailingslashit( $this->namespace ) . 'settings/' . $partial->primary_setting ),
+				'embeddable' => true,
+				),
+			);
+		}
+
+		foreach ( $partial as $property => $value ) {
+			if ( in_array( $property, $hide_from_response, true ) ) {
+				continue;
+			} elseif ( 'settings' === $property ) {
```

Comment:
I think this condition could be extracted from this `foreach` loop, and instead put it above in a combined loop with the `primary_setting` handling. For example:

```php
$settings = array_unique( array_filter( array_merge( array( $partial->primary_setting ), $partial->settings ) ) );
foreach ( $settings ) {
    // ...
}
```